hpx/hpx_distributed.cc

```diff
@@ -0,0 +1,369 @@
+                                                        output.end());
+                    } else {
+                      auto output_new = outputs_new[dep - first_point];
+                      point_inputs[point_n_inputs].assign(output_new.begin(),
+                                                        output_new.end());
+                    }
+                  } else {
+                    int from = tag_bits_by_point[dep];
+                    int to = tag_bits_by_point[point];
+                    int tag = (from << 8) | to;
+                    MPI_Request req;
+                    MPI_Irecv(point_inputs[point_n_inputs].data(),
+                              point_inputs[point_n_inputs].size(), MPI_BYTE,
+                              rank_by_point[dep], tag, MPI_COMM_WORLD, &req);
+                    requests.push_back(req);
+                  }
+                  point_n_inputs++;
+                }
+              }
+            }  // receive
+
+            // Send
+            if (point >= last_offset && point < last_offset + last_width) {
+              for (auto interval : point_rev_deps) {
+                for (long dep = interval.first; dep <= interval.second; dep++) {
+                  if (dep < offset || dep >= offset + width ||
+                      (first_point <= dep && dep <= last_point)) {
+                    continue;
+                  }
+                  int from = tag_bits_by_point[point];
+                  int to = tag_bits_by_point[dep];
+                  int tag = (from << 8) | to;
+                  MPI_Request req;
+                  if (timestep % 2 == 0) {
+                      MPI_Isend(point_output.data(), point_output.size(), MPI_BYTE,
+                            rank_by_point[dep], tag, MPI_COMM_WORLD, &req);
+                  } else {
+                      MPI_Isend(point_output_new.data(), point_output_new.size(), MPI_BYTE,
+                            rank_by_point[dep], tag, MPI_COMM_WORLD, &req);
+                  }
+                  requests.push_back(req);
+                }
+              }
+            }  // send
+
+            MPI_Waitall(requests.size(), requests.data(), MPI_STATUSES_IGNORE);
+
+            if (timestep % 2 == 0) {
+                graph.execute_point(timestep, point, point_output_new.data(),
+                                point_output_new.size(), point_input_ptr.data(),
+                                point_input_bytes.data(), point_n_inputs,
+                                scratch_ptr + scratch_bytes * point_index,
+                                scratch_bytes);
+            } else {
+                graph.execute_point(timestep, point, point_output.data(),
+                                point_output.size(), point_input_ptr.data(),
+                                point_input_bytes.data(), point_n_inputs,
+                                scratch_ptr + scratch_bytes * point_index,
+                                scratch_bytes);
+            }
```

Comment:
Instead of having all these mod checks, I wonder if it would be clearer to have a single `std::swap` call at the end? Should still be efficient.

Reply:
I dont think use std::swap(point_output, point_output_new) could work here. The reason is, the communication (send and receive) and execute kernel is in same hpx for loop. For example, we have stencil pattern on a node with 48 cores, and we have a task of 48 width. core 0 will need to send to core 1 (use on-node sharing); core 1 will need to send to core 0 and core 2 (use on-node sharing); ... If we use std::swap here, core 1 executes the kernel and generates the new output and may send this new output to core 0 and core 2. But core 0 and core 2 are expecting the previous output. So I am creating two variable to store the output, one is to store the output needed for current time step, another is to store  the new generated output needed for next time step.